src/Common.OData.ApiExplorer/AspNet.OData/DefaultModelTypeBuilder.cs

```diff
@@ -30,6 +30,7 @@ public sealed class DefaultModelTypeBuilder : IModelTypeBuilder
         readonly ICollection<Assembly> assemblies;
         readonly ConcurrentDictionary<ApiVersion, ModuleBuilder> modules = new ConcurrentDictionary<ApiVersion, ModuleBuilder>();
         readonly ConcurrentDictionary<ClassSignature, Type> generatedTypes = new ConcurrentDictionary<ClassSignature, Type>();
+        private readonly Dictionary<string, Type> passedTypes = new Dictionary<string, Type>();
```

Comment:
In thinking about this, I'm pretty sure that using `string` alone for the key is not going to work because there can be a distinctly generated type by API version. This should mean that the key for the corresponding EDM type should be a combination of EDM type and API version; otherwise, it's possible that the wrong generated type will be picked.

For example, imagine that `Default.Company` is generated for API version 2.0 and then it's requested again for API version 1.0. If the models aren't exactly the same (which they probably aren't), the wrong selection is made.

There's two ways this can be handled:
1. Add a hashable struct, something like **VisitedTypeKey**
2. Use `int` as the key and add a hashing function

Option 2 might look like:

```c#
static int GetVisitedTypeKey( IEdmTypeReference type, ApiVersion apiVersion ) =>
  ( type.FullName().GetHashCode() * 397 ) ^ apiVersion.GetHashCode();
```